File: src/commands/sync.ts
import { Command } from "commander";
import chalk from "chalk";
import { WorktreeManager } from "../git/WorktreeManager";
import { handleCommandError } from "../utils";

interface SyncCommandOptions {
  branch?: string;
}

export function createSyncCommand(): Command {
  const command = new Command("sync");

  command
    .description("Sync the bare clone with the latest changes from origin")
    .option(
      "-b, --branch <branch>",
      "Branch to sync (defaults to main or master)",
    )
    .action(async (options: SyncCommandOptions) => {
      try {
        await runSync(options);
      } catch (error) {
        handleCommandError(error);
      }
    });

  return command;
}

async function runSync(options: SyncCommandOptions): Promise<void> {
  // Use discovery to find the bare clone from anywhere in the project hierarchy
  const manager = await WorktreeManager.discover();

  // Determine the branch to sync
  const branch = options.branch || (await manager.getDefaultBranch());

  // Check if the branch is checked out in any worktree
  const worktrees = await manager.listWorktrees();
  const checkedOutWorktree = worktrees.find((wt) => wt.branch === branch);

  if (checkedOutWorktree) {
    console.log(
      chalk.yellow(`Branch '${branch}' is checked out in a worktree. Git won't sync against a checked-out branch (a limitation of git worktrees).`),
    );
    console.log(
      chalk.yellow(`Run 'git fetch origin', then merge or rebase from 'origin/${branch}'.`),
    );
    process.exit(1);
  }

  await manager.syncBranch(branch);

  console.log(chalk.green("✓ Synced"), chalk.bold(branch), chalk.gray("from origin"));
}

File: src/commands/sync.rs
use colored::Colorize;

use crate::git::{discover_repo, get_default_branch, list_worktrees, sync_branch};

pub fn run(branch: Option<&str>) {
    let repo = match discover_repo() {
        Ok(m) => m,
        Err(e) => {
            eprintln!("{} {}", "Error:".red(), e);
            std::process::exit(1);
        }
    };

    let target_branch = if let Some(b) = branch {
        b.to_string()
    } else {
        match get_default_branch(&repo) {
            Ok(b) => b,
            Err(e) => {
                eprintln!("{} {}", "Error:".red(), e);
                std::process::exit(1);
            }
        }
    };

    // Check if the branch is checked out in any worktree
    if let Ok(worktrees) = list_worktrees(&repo) {
        if worktrees.iter().any(|wt| wt.branch == target_branch) {
            println!(
                "{}",
                format!(
                    "Branch '{}' is checked out in a worktree. Git won't sync against a checked-out branch (a limitation of git worktrees).",
                    target_branch
                )
                .yellow()
            );
            println!(
                "{}",
                format!(
                    "Run 'git fetch origin', then merge or rebase from 'origin/{}'.",
                    target_branch
                )
                .yellow()
            );
            std::process::exit(1);
        }
    }

    if let Err(e) = sync_branch(&repo, &target_branch) {
        eprintln!("{} {}", "Error:".red(), e);
        std::process::exit(1);
    }

    println!(
        "{} {} {}",
        "✓ Synced".green(),
        target_branch.bold(),
        "from origin".dimmed()
    );
}
